### PR TITLE
GraphQL vector updates

### PR DESCRIPTION
**Pull request by ricopinazo**
- **New vector API (#1717)**
- **wip**

### Issues

_If this resolves any issues, please link to them here, the format is a KEYWORD followed by @<ISSUE NUMBER>__
KEYWORDS available are `close`, `closes`, `closed`, `fix`, `fixes`, `fixed`, `resolve`, `resolves`, `resolved`.
Please delete this text before creating your PR

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -77,6 +77,7 @@ thiserror = "1.0.57"
 dotenv = "0.15.0"
 csv = "1.3.0"
 flate2 = "1.0.28"
+tar = "0.4.41"
```
Comment (ricopinazo) [inline Cargo.toml:80]: remove this

File: python/tests/graphql/edit_graph/test_graphql.py [hunk 2]
```diff
@@ -138,38 +138,38 @@ def assert_graph_fetch(path):
         path = "../shivam/g"
         with pytest.raises(Exception) as excinfo:
             client.send_graph(path=path, graph=g, overwrite=True)
-        assert "References to the parent dir are not allowed within the path:" in str(
+        assert "Invalid path: ParentDirNotAllowed" in str(
```
Comment (ricopinazo) [inline python/tests/graphql/edit_graph/test_graphql.py:141]: have a look at this

File: raphtory-graphql/src/data.rs [hunk 3]
```diff
@@ -1,349 +1,251 @@
...
+    pub(crate) global_template: Option<DocumentTemplate>,
+    pub(crate) individual_templates: HashMap<PathBuf, DocumentTemplate>,
+}
+
+#[derive(Clone)]
 pub struct Data {
-    pub(crate) work_dir: PathBuf,
-    pub(crate) graphs: Cache<PathBuf, IndexedGraph<MaterializedGraph>>,
-    pub(crate) global_plugins: GlobalPlugins,
+    pub(crate) work_dir: PathBuf, // TODO: make private again!!!
```
Comment (ricopinazo) [inline raphtory-graphql/src/data.rs:37]: review this

File: raphtory-graphql/src/data.rs [hunk 4]
```diff
@@ -1,349 +1,251 @@
...
-        let graphs_cache_builder = Cache::<_, IndexedGraph<MaterializedGraph>>::builder()
+        let cache = Cache::<PathBuf, GraphWithVectors>::builder()
             .max_capacity(cache_configs.capacity)
             .time_to_idle(std::time::Duration::from_secs(cache_configs.tti_seconds))
             .eviction_listener(|_, value, _| {
                 value
+                    .graph
                     .write_updates()
                     .unwrap_or_else(|err| error!("Write on eviction failed: {err:?}"))
+                // FIXME: don't have currently a way to know which embedding updates are pending
```
Comment (ricopinazo) [inline raphtory-graphql/src/data.rs:54]: dump all the embeddings

File: raphtory-graphql/src/data.rs [hunk 5]
```diff
@@ -1,349 +1,251 @@
...
-    Ok(graph)
-}
-
-#[allow(unused_variables)]
-#[cfg(not(feature = "storage"))]
-fn _load_disk_graph(_path: &Path) -> Result<MaterializedGraph, GraphError> {
-    unimplemented!("Storage feature not enabled, cannot load from disk graph")
 }
 
+// TODO: review if we are using this
```
Comment (ricopinazo) [inline raphtory-graphql/src/data.rs:223]: remove this

File: raphtory-graphql/src/data.rs [hunk 6]
```diff
@@ -354,8 +256,8 @@ pub(crate) mod data_tests {
     };
 
     use crate::config::app_config::{AppConfig, AppConfigBuilder};
-    #[cfg(feature = "storage")]
-    use crate::data::copy_dir_recursive;
+    // #[cfg(feature = "storage")]
```
Comment (ricopinazo) [inline raphtory-graphql/src/data.rs:259]: remove this

File: raphtory-graphql/src/data.rs [hunk 7]
```diff
@@ -399,27 +302,25 @@ pub(crate) mod data_tests {
         Ok(())
     }
 
+    // TODO: this is only used for testing, should be under some test flag
     pub(crate) fn save_graphs_to_work_dir(
```
Comment (ricopinazo) [inline raphtory-graphql/src/data.rs:306]: this is only used for tests

File: raphtory-graphql/src/graph.rs [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+                    .folder
+                    .get()
+                    .ok_or(GraphError::CacheNotInnitialised)?
+                    .get_vectors_path(),
+            );
+        }
+        Ok(())
+    }
+
+    pub(crate) fn read_from_folder(
```
Comment (ricopinazo) [inline raphtory-graphql/src/graph.rs:92]: test reading and writing the entire thing

File: raphtory-graphql/src/model/mod.rs [hunk 9]
```diff
@@ -160,217 +144,70 @@ impl Mut {
...
-            new_graph.update_constant_properties([("lastOpened", Prop::I64(timestamp * 1000))])?;
-            create_dirs_if_not_present(&new_full_path)?;
-            new_graph.encode(&new_full_path)?;
-        }
+        // TODO: review, can't we do here as well:
+        // if graph.graph.storage().is_immutable() {
+        //     return Err(GqlGraphError::ImmutableDiskGraph.into());
+        // }
+        let timestamp: i64 = Utc::now().timestamp();
+        graph.update_constant_properties([("lastUpdated", Prop::I64(timestamp * 1000))])?;
```
Comment (ricopinazo) [inline raphtory-graphql/src/model/mod.rs:179]: I should remove this

File: raphtory-graphql/src/model/mod.rs [hunk 10]
```diff
@@ -402,21 +239,20 @@ impl Mut {
...
         }
-
-        let mut in_file = graph.value(ctx)?.content;
-        create_dirs_if_not_present(&full_path)?;
-        let mut out_file = File::create(&full_path)?;
-        copy(&mut in_file, &mut out_file)?;
-        let g = MaterializedGraph::load_cached(&full_path)?;
-        data.graphs.insert(path.to_path_buf(), g.into());
-        Ok(path.display().to_string())
+        data.insert_graph(&path, graph).await?;
```
Comment (ricopinazo) [inline raphtory-graphql/src/model/mod.rs:254]: create a ticket for optimizing this and also the dump to disk for the vectors

File: raphtory-graphql/src/paths.rs [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+
+impl Deref for ValidGraphFolder {
+    type Target = GraphFolder;
+
+    fn deref(&self) -> &Self::Target {
+        &self.folder
+    }
+}
+
+impl ValidGraphFolder {
```
Comment (ricopinazo) [inline raphtory-graphql/src/paths.rs:84]: add the .raph file as empty for now

File: raphtory-graphql/src/python/global_plugins.rs [hunk 12]
```diff
@@ -49,31 +51,27 @@ impl PyGlobalPlugins {
         window: PyWindow,
     ) -> Vec<(PyDocument, f32)> {
         let window = translate_window(window);
-        let graphs = self.0.vectorised_graphs.read();
+        let graphs = &self.0.graphs;
         let cluster = VectorisedCluster::new(&graphs);
-        let vectorised_graphs = self.0.vectorised_graphs.read();
-        let graph_entry = vectorised_graphs.iter().next();
+        // let vectorised_graphs = self.0.vectorised_graphs.read();
```
Comment (ricopinazo) [inline raphtory-graphql/src/python/global_plugins.rs:56]: remove this comment

File: raphtory-graphql/src/server.rs [hunk 13]
```diff
@@ -100,6 +97,23 @@ impl GraphServer {
         Ok(Self { data, config })
     }
 
+    pub fn set_embeddings<F: EmbeddingFunction + Clone + 'static>(
+        mut self,
+        embedding: F, // TODO: make this optional
+        cache: &Path, // TODO: maybe now that we are storing vectors we could bin the cache!!!
```
Comment (ricopinazo) [inline raphtory-graphql/src/server.rs:103]: remove this TODOS

File: raphtory/src/graph_loader/stable_coins.rs [hunk 14]
```diff
@@ -62,7 +62,7 @@ pub fn stable_coin_graph(path: Option<String>, subset: bool) -> Graph {
         }
     }
 
-    let encoded_data_dir = data_dir.join("graphdb.bincode");
+    let encoded_data_dir = data_dir.join("graphdb.bincode"); // this as a dir in windows might fail ?
```
[the remark this answers is not in the document]
Comment (ricopinazo) [reply raphtory/src/graph_loader/stable_coins.rs:65]: remove dot from there

File: raphtory/src/serialise/mod.rs [hunk 15]
```diff
@@ -6,5 +9,137 @@ mod proto {
...
+    }
+
+    pub fn write_graph(&self, buf: &[u8]) -> Result<(), io::Error> {
+        if self.prefer_zip_format {
+            let file = File::create(&self.path)?;
+            let mut zip = ZipWriter::new(file);
+            zip.start_file::<_, ()>("graph", FileOptions::default())?;
+            zip.write_all(buf)
+        } else {
+            let _ignored = fs::create_dir(&self.path);
```
Comment (ricopinazo) [inline raphtory/src/serialise/mod.rs:77]: replace with _

File: raphtory/src/serialise/mod.rs [hunk 16]
```diff
@@ -6,5 +9,137 @@ mod proto {
...
+    io::{self, Read, Write},
+    path::{Path, PathBuf},
+};
+
 pub use proto::Graph as ProtoGraph;
 pub use serialise::{CacheOps, StableDecode, StableEncode};
+
+#[derive(Clone, Debug)]
+pub struct GraphFolder {
+    path: PathBuf,
```
Comment (ricopinazo) [inline raphtory/src/serialise/mod.rs:23]: rename this to base_path

File: raphtory/src/serialise/mod.rs [hunk 17]
```diff
@@ -6,5 +9,137 @@ mod proto {
...
+            zip.write_all(buf)
+        } else {
+            let _ignored = fs::create_dir(&self.path);
+            let mut file = File::create(self.get_graph_path())?;
+            file.write_all(buf)
+        }
+    }
+
+    fn get_appendable_graph_file(&self) -> Result<File, std::io::Error> {
+        let _ignored = fs::create_dir(&self.path);
```
Comment (ricopinazo) [inline raphtory/src/serialise/mod.rs:84]: the same here

File: raphtory/src/serialise/mod.rs [hunk 18]
```diff
@@ -6,5 +9,137 @@ mod proto {
...
+        }
+    }
+
+    pub fn write_graph(&self, buf: &[u8]) -> Result<(), io::Error> {
+        if self.prefer_zip_format {
+            let file = File::create(&self.path)?;
+            let mut zip = ZipWriter::new(file);
+            zip.start_file::<_, ()>("graph", FileOptions::default())?;
+            zip.write_all(buf)
+        } else {
```
Comment (ricopinazo) [inline raphtory/src/serialise/mod.rs:76]: if the directory is empty is fine for writing a graph. For vectors we don't check

File: raphtory/src/vectors/vector_storage.rs [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+            template,
+            embedding,
+            cache_storage,
+            Arc::new(graph_documents.into()),
+            Arc::new(node_documents.into()),
+            Arc::new(edge_documents.into()),
+        ))
+    }
+
+    pub fn write_to_path(&self, path: &Path) {
```
Comment (ricopinazo) [inline raphtory/src/vectors/vector_storage.rs:54]: return an error here

File: raphtory-graphql/src/graph.rs [hunk 20]
```diff
@@ -0,0 +1,165 @@
...
+        let vectors = VectorisedGraph::read_from_path(
+            &folder.get_vectors_path(),
+            graph.clone(),
+            embedding,
+            cache,
+        );
+
+        println!("Graph loaded = {}", folder.get_original_path_str());
+        Ok(Self {
+            graph: IndexedGraph::from_graph(&graph)?,
```
Comment (miratepuffin) [inline raphtory-graphql/src/graph.rs:113]: We need to test how long the indexing is taking on big graphs

File: raphtory-graphql/src/model/graph/mutable_graph.rs [hunk 21]
```diff
@@ -173,16 +165,17 @@ impl GqlMutableGraph {
                 )?;
             }
             let constant_props = edge.constant_properties.unwrap_or(vec![]);
+            let edge_view = self
```
Comment (miratepuffin) [inline raphtory-graphql/src/model/graph/mutable_graph.rs:168]: This needs to go back inside - if the vec of updates is empty, no need to update embeddings

File: raphtory-graphql/src/python/server/server.rs [hunk 22]
```diff
@@ -182,81 +192,75 @@ impl PyGraphServer {
...
-    ///   cache (str):  the directory to use as cache for the embeddings.
-    ///   embedding (Function):  the embedding function to translate documents to embeddings.
-    ///   graph_template (String):  the template to use for graphs.
-    ///   node_template (String):  the template to use for nodes.
-    ///   edge_template (String):  the template to use for edges.
-    ///
-    /// Returns:
-    ///    GraphServer: A new server object containing the vectorised graphs.
-    fn with_vectorised(
+    #[pyo3(
```
Comment (miratepuffin) [inline raphtory-graphql/src/python/server/server.rs:195]: can you quickly add some basic docs

File: raphtory-graphql/src/server.rs [hunk 23]
```diff
@@ -186,6 +167,8 @@ impl GraphServer {
 
     /// Start the server on the port `port` and return a handle to it.
     pub async fn start_with_port(self, port: u16) -> IoResult<RunningGraphServer> {
+        self.data.vectorise_all_graphs_that_are_not().await; // is it ok ignoring this error?
```
Comment (miratepuffin) [inline raphtory-graphql/src/server.rs:170]: Can we not ignore please

**Review by miratepuffin — COMMENTED** [review]
Couple small changes and we are good to go!

File: raphtory-graphql/src/model/graph/mutable_graph.rs [hunk 24]
```diff
@@ -111,21 +105,18 @@ impl GqlMutableGraph {
                     None,
                 )?;
             }
+            let node_view = self
```
[the remark this answers is not in the document]
Comment (ricopinazo) [reply raphtory-graphql/src/model/graph/mutable_graph.rs:108]: Done

**Review by miratepuffin — APPROVED** [review]
LGTM

**Pull request merged**